packages/rpc-methods/src/permitted/invokeKeyring.ts

```diff
@@ -0,0 +1,141 @@
+import type { Snap } from '@metamask/snaps-utils';
+import {
+  HandlerType,
+  WALLET_SNAP_PERMISSION_KEY,
+  type SnapId,
+  type SnapRpcHookArgs,
+} from '@metamask/snaps-utils';
+import type {
+  PermittedHandlerExport,
+  PendingJsonRpcResponse,
+  JsonRpcEngineEndCallback,
+  JsonRpcRequest,
+} from '@metamask/types';
+import { hasProperty, type Json } from '@metamask/utils';
+import { ethErrors } from 'eth-rpc-errors';
+
+import type { MethodHooksObject } from '../utils';
+import type { InvokeSnapSugarArgs } from './invokeSnapSugar';
+import { getValidatedParams } from './invokeSnapSugar';
+
+const hookNames: MethodHooksObject<InvokeKeyringHooks> = {
+  hasPermission: true,
+  handleSnapRpcRequest: true,
+  getSnap: true,
+  getAllowedKeyringMethods: true,
+};
+
+/**
+ * `wallet_invokeKeyring` gets the requester's permitted and installed Snaps.
+ */
+export const invokeKeyringHandler: PermittedHandlerExport<
+  InvokeKeyringHooks,
+  JsonRpcRequest<unknown>,
+  unknown
+> = {
+  methodNames: ['wallet_invokeKeyring'],
+  implementation: invokeKeyringImplementation,
+  hookNames,
+};
+
+export type InvokeKeyringHooks = {
+  hasPermission: (origin: string, permissionName: string) => boolean;
+
+  handleSnapRpcRequest: ({
```

Comment:
I think at some point we should change these hooks (same for the other methods) to do the bare minimum needed for the method to work. This one could just be calling `onKeyringRequest` with the already known origin for example, so we don't specify the origin and handler here.

That's something we can change later though.